Added migration guide (#13)
openai/openai-dotnet: MigrationGuide.md

> # Guide for migrating to OpenAI 2.0.0-beta.1 from OpenAI 1.11.0

Reviewer: Would it be neater to move the migration guide under a "docs" folder instead of the repo's root? 🤔

Author: We could. I left it besides the README and CHANGELOG because that's what I usually see in the .NET repo. We don't seem to get too many docs files other than the migration guide and a possible troubleshooting guide.

This guide is intended to assist in the migration to the official OpenAI library (2.0.0-beta.1) from [OpenAI 1.11.0][openai_1110], focusing on side-by-side comparisons for similar operations between libraries. In this document, these are often referred to as the 'new' and the 'old' OpenAI libraries, respectively.

> 

Reviewer 2: I'd recommend an explicit mention that, prior to 2.0, the `OpenAI` package was a community library not officially by OpenAI and that it's now the official library. Linking the recent changelog update could likely facilitate that, too!

> Familiarity with the OpenAI 1.11.0 package is assumed. For those new to any OpenAI SDK for .NET, please refer to the [README][readme] rather than this guide.

Reviewer 3: ```suggestion
Familiarity with the OpenAI 1.11.0 package is assumed. For those new to any OpenAI library for .NET, see the [README][readme] rather than this guide.
```

## Table of contents
- [Client usage](#client-usage)
- [Authentication](#authentication)
- [Highlighted scenarios](#highlighted-scenarios)

>     - [Chat Completions: text generation](#chat-completions-text-generation)
>     - [Chat Completions: streaming](#chat-completions-streaming)
>     - [Chat Completions: JSON mode](#chat-completions-json-mode)
>     - [Chat Completions: vision](#chat-completions-vision)
>     - [Audio: speech-to-text](#audio-speech-to-text)
>     - [Audio: text-to-speech](#audio-text-to-speech)
>     - [Image: image generation](#image-image-generation)

Reviewer 3: ```suggestion
    - [Chat Completions: Text generation](#chat-completions-text-generation)
    - [Chat Completions: Streaming](#chat-completions-streaming)
    - [Chat Completions: JSON mode](#chat-completions-json-mode)
    - [Chat Completions: Vision](#chat-completions-vision)
    - [Audio: Speech-to-text](#audio-speech-to-text)
    - [Audio: Text-to-speech](#audio-text-to-speech)
    - [Image: Image generation](#image-image-generation)
```

> - [Additional samples](#additional-samples)

Reviewer: nit: Replace "samples" for "examples" everywhere, for consistency.

Old library:
```cs
OpenAIAPI api = new OpenAIAPI("<api-key>");
ImageResult result = await api.ImageGenerations.CreateImageAsync("Draw a quick brown fox jumping over a lazy dog.", Model.DALLE3);
```

New library:
```cs
ImageClient client = new ImageClient("dall-e-3", "<api-key>");
ClientResult<GeneratedImage> result = await client.GenerateImageAsync("Draw a quick brown fox jumping over a lazy dog.");
```

Another major difference highlighted in the snippets above is that the new library requires the model to be explicitly set during client instantiation, while the old client allows a model to be specified per call.

The table below illustrates to which client each endpoint of `OpenAIAPI` was ported. Note that the deprecated Completions API is not supported by the new library:

Old library's endpoint|New library's client
|-|-
|Chat | ChatClient
|ImageGenerations | ImageClient
|TextToSpeech | AudioClient
|Transcriptions | AudioClient
|Translations | AudioClient
|Moderation | ModerationClient
|Embeddings | EmbeddingClient
|Files | FileClient
|Models | ModelClient
|Completions | Not supported

## Authentication

> In order to authenticate to OpenAI you must set an API key when creating a client.

Reviewer 3: ```suggestion
To authenticate to OpenAI, you must set an API key when creating a client.
```

> The old library allowed setting the API key in three different ways: directly from a string, from an environment variable, or from a configuration file.

Reviewer 3: ```suggestion
The old library allowed setting the API key in 3 different ways:
- Directly from a string
- From an environment variable
- From a configuration file
```

Author: Couldn't commit the suggested changes because it got outdated. Just updated it manually.

// Sets the API key directly from a string.
api = new OpenAIAPI("<api-key>");

// Attempts to load the API key from environment variables OPENAI_KEY and OPENAI_API_KEY.
api = new OpenAIAPI(APIAuthentication.LoadFromEnv());

// Attempts to load the API key from a configuration file.
api = new OpenAIAPI(APIAuthentication.LoadFromPath("<directory>", "<filename>"));
```

OpenAI 2.0.0-beta.1 only supports setting it from a string or from an environment variable. The following snippet illustrates the behavior with the `ChatClient`, but other clients behave the same:

```cs
ChatClient client;

// Sets the API key directly from a string.
client = new ChatClient("gpt-3.5-turbo", "<api-key>");

// When no API key string is specified, attempts to load the API key from the environment variable OPENAI_API_KEY.
client = new ChatClient("gpt-3.5-turbo");
```

Note that, unlike the OpenAI 1.11.0, OpenAI 2.0.0-beta.1 will never attempt to load the API key from the `OPENAI_KEY` environment variable. Only `OPENAI_API_KEY` is supported.

## Highlighted scenarios

The following sections illustrate side-by-side comparisons for similar operations between the two libraries, highlighting common scenarios.

> ### Chat Completions: text generation

Reviewer 3: ```suggestion
### Chat Completions: Text generation
```

conversation.AppendUserInput("When was the Nobel Prize founded?");

await conversation.GetResponseFromChatbotAsync();

conversation.AppendUserInput("Who was the first person to be awarded one?");

await conversation.GetResponseFromChatbotAsync();

foreach (ChatMessage message in conversation.Messages)
{
    Console.WriteLine($"{message.Role}: {message.TextContent}");
}
```

OpenAI 2.0.0-beta.1:
```cs
ChatClient client = new ChatClient("gpt-3.5-turbo", "<api-key>");
List<ChatMessage> messages = new List<ChatMessage>()
{
    new SystemChatMessage("You are a helpful assistant."),
    new UserChatMessage("When was the Nobel Prize founded?")
};

ClientResult<ChatCompletion> result = await client.CompleteChatAsync(messages);

messages.Add(new AssistantChatMessage(result));
messages.Add(new UserChatMessage("Who was the first person to be awarded one?"));

result = await client.CompleteChatAsync(messages);

messages.Add(new AssistantChatMessage(result));

foreach (ChatMessage message in messages)
{
    string role = message.GetType().Name;
    string text = message.Content[0].Text;

    Console.WriteLine($"{role}: {text}");
}
```

> ### Chat Completions: streaming

Reviewer 3: ```suggestion
### Chat Completions: Streaming
```

OpenAI 1.11.0:
```cs
OpenAIAPI api = new OpenAIAPI("<api-key>");
ChatRequest request = new ChatRequest()
{
    Model = Model.ChatGPTTurbo,
    ResponseFormat = request.ResponseFormats.JsonObject,
    Messages = new List<ChatMessage>()
    {
        new ChatMessage(ChatMessageRole.System, "You are a helpful assistant designed to output JSON."),
        new ChatMessage(ChatMessageRole.User, "Give me a JSON object listing Nobel Prize winners of the last 5 years.")
    }
};

ChatResult result = await api.Chat.CreateChatCompletionAsync(request);

Console.WriteLine(result);
```

OpenAI 2.0.0-beta.1:
```cs
ChatClient client = new ChatClient("gpt-3.5-turbo", "<api-key>");
List<ChatMessage> messages = new List<ChatMessage>()
{
    new SystemChatMessage("You are a helpful assistant designed to output JSON."),
    new UserChatMessage("Give me a JSON object listing Nobel Prize winners of the last 5 years.")
};
ChatCompletionOptions options = new ChatCompletionOptions()
{
    ResponseFormat = ChatResponseFormat.JsonObject
};

ClientResult<ChatCompletion> result = await client.CompleteChatAsync(messages, options);
string text = result.Value.Content[0].Text;

Console.WriteLine(text);
```

> ### Chat Completions: vision

Reviewer 3: ```suggestion
### Chat Completions: Vision
```

OpenAI 1.11.0:
```cs
OpenAIAPI api = new OpenAIAPI("<api-key>");
Conversation conversation = api.Chat.CreateConversation();
byte[] imageData = await File.ReadAllBytesAsync("<file-path>");

conversation.Model = Model.GPT4_Vision;
conversation.AppendUserInput("Describe this image.", ImageInput.FromImageBytes(imageData));

string response = await conversation.GetResponseFromChatbotAsync();

Console.WriteLine(response);
```

OpenAI 2.0.0-beta.1:
```cs
ChatClient client = new ChatClient("gpt-4-vision-preview", "<api-key>");
using FileStream file = File.OpenRead("<file-path>");
BinaryData imageData = await BinaryData.FromStreamAsync(file);
List<ChatMessage> messages = new List<ChatMessage>()
{
    new UserChatMessage(
        ChatMessageContentPart.CreateTextMessageContentPart("Describe this image."),
        ChatMessageContentPart.CreateImageMessageContentPart(imageData, "image/png"))
};

ClientResult<ChatCompletion> result = await client.CompleteChatAsync(messages);
string text = result.Value.Content[0].Text;

Console.WriteLine(text);
```

> ### Audio: speech-to-text

Reviewer 3: ```suggestion
### Audio: Speech-to-text
```

OpenAI 1.11.0:
```cs
OpenAIAPI api = new OpenAIAPI("<api-key>");
string result = await api.Transcriptions.GetTextAsync("<file-path>", "fr");

Console.WriteLine(result);
```

OpenAI 2.0.0-beta.1:
```cs
AudioClient client = new AudioClient("whisper-1", "<api-key>");
AudioTranscriptionOptions options = new AudioTranscriptionOptions()
{
    Language = "fr"
};

ClientResult<AudioTranscription> result = await client.TranscribeAudioAsync("<file-path>", options);
string text = result.Value.Text;

Console.WriteLine(text);
```

> ### Audio: text-to-speech

Reviewer 3: ```suggestion
### Audio: Text-to-speech
```

OpenAI 1.11.0:
```cs
OpenAIAPI api = new OpenAIAPI("<api-key>");
TextToSpeechRequest request = new TextToSpeechRequest()
{
    Input = "Hasta la vista, baby.",
    Model = Model.TTS_Speed,
    Voice = "alloy"
};

await api.TextToSpeech.SaveSpeechToFileAsync(request, "<file-path>");
```

OpenAI 2.0.0-beta.1:
```cs
AudioClient client = new AudioClient("tts-1", "<api-key>");

ClientResult<BinaryData> result = await client.GenerateSpeechFromTextAsync("Hasta la vista, baby.", GeneratedSpeechVoice.Alloy);
BinaryData data = result.Value;

await File.WriteAllBytesAsync("<file-path>", data.ToArray());
```

> ### Image: image generation

Reviewer 3: ```suggestion
### Image: Image generation
```
